PROGRESS.md

```diff
@@ -0,0 +1,406 @@
+task a {
+  command { ... }
+  output { Int o = read_int(stdout()) }
+}
+
+task b {
+  Array[Int] ints
+  command { ... }
+  output { Int o = read_int(stdout()) }
+}
+
+workflow w {
+  call a as a0
+  call a as a1
+  Array[Int] ints = [a0.o, a1.o]
+  call b {input: ints=ints}
+}
+```
+
+`b.upstream` evaluates to a set of one element: the declaration for `ints` in the workflow.  `upstream` and `downstream` only return `GraphNode`s that they *directly depend on*.
+
+If you'd like `b.upstream` to return `call a as a0`, `call a as a1`, AND `Array[Int] ints = [a0.o, a1.o]`, this can be done easily like this:
+
+```scala
+def upstreamRecursive(node: GraphNode): Set[Scope with GraphNode] = {
+  node.upstream ++ node.upstream.flatMap(upstreamRecursive)
+}
+```
+
+### Lexical Relationships
+
+A lexical relationship describes `Scope`s relate to each other.  This relationship is always tied to how they are represented in source code.  Currently the following classes are `Scope`s:
+
+* `Call`
+* `Declaration`
+* `Task`
+* `Workflow`
+* `Namespace`
+* `If`
+* `Scatter`
+
+Lexical operations:
+
+```scala
+trait Scope {
+  def parent: Option[Scope]
+  def children: Seq[Scope] // in source code order
+  def ancestry: Seq[Scope]
+  def descendants: Set[Scope]
+  def closestCommonAncestor(other: Scope): Option[Scope]
+  def resolveVariable(name: String): Option[Scope with GraphNode]
+  def lookupFunction(inputs: WorkflowCoercedInputs,
+                     wdlFunctions: WdlFunctions[WdlValue],
+                     shards: Map[Scatter, Int] = Map.empty[Scatter, Int],
+                     relativeTo: Scope = this): String => WdlValue
+}
+```
+
+
+
+If statements
+=============
+
+```
+task A {
+  String i
+  command {ps}
+  output {String o = read_string(stdout())}
+}
+
+workflow w {
+  Int i
+  Array[String] arr
+
+  call A
+
+  if (i == 2) {
+    call A as B
+  }
+
+  if (A.o == "foo") {
+    call A as C
+  }
+
+  if (A.o == "bar") {
+    scatter(x in arr) {
+      call A as D {input: i=x}
+    }
+  }
+
+  call A as E {input: i=C.o}
+}
+```
+
+For the above WDL:
+
+```scala
+val d = namespace.calls.find(_.unqualifiedName == "D").get
+
+println(d.ancestry.map(_.fullyQualifiedNameWithIndexScopes))
+// List(w.$if_2.$scatter_0, w.$if_2, w, )
+
+println(d.upstream.map(_.fullyQualifiedNameWithIndexScopes))
+// Set(w.$if_2.$scatter_0)
+
+println(upstreamRecursive(d).map(_.fullyQualifiedNameWithIndexScopes))
+//Set(w.$if_2.$scatter_0, w.arr, w.$if_2, w.A)
+```
+
+Nested Scatters
+===============
+
+```
+task inner {
+  Int i
+  command { echo ${i} }
+  output {
+    Int out = read_int(stdout()) + 1
+  }
+}
+
+task outer {
+  Array[Array[Int]] matrix
+  command { cat ${write_tsv(matrix)} }
+  output {
+    String tsv = read_string(stdout())
+  }
+}
+
+workflow w {
+  Array[Array[Int]] array = [[0,1,2],[3,4,5],[6,7,8]]
+
+  scatter(i in array) {
+    scatter(j in i) {
+      call inner {input: i=j}
+    }
+  }
+
+  # inner.out should be [[1,2,3],[4,5,6],[7,8,9]] (???)
```

Comment:
I'm guessing that's Scott's way of wondering if this should be the output for inner.out when we implement nested scatters.